crafters/image/ImageReader/tests/test_imagereader.py

```diff
@@ -3,35 +3,95 @@
 
 import numpy as np
 from PIL import Image
+import pytest
 
 from .. import ImageReader
 
 
-def create_test_image(output_fn, size_width=50, size_height=50):
+def create_pil_image(size_width, size_height):
     from PIL import Image
-    image = Image.new('RGB', size=(size_width, size_height), color=(155, 0, 0))
-    with open(output_fn, "wb") as f:
+    image = Image.new('RGB', size=(size_width, size_height), color=(50, 100, 150))
+    return image
+
+
+def create_test_image(filename, size_width, size_height):
+    image = create_pil_image(size_width, size_height)
+    with open(filename, "wb") as f:
         image.save(f, 'jpeg')
 
 
 def test_io_uri():
     crafter = ImageReader()
-    tmp_fn = os.path.join(crafter.current_workspace, 'test.jpeg')
-    img_size = 50
-    create_test_image(tmp_fn, size_width=img_size, size_height=img_size)
-    test_doc = crafter.craft(buffer=None, uri=tmp_fn)
-    assert test_doc['blob'].shape == (img_size, img_size, 3)
+    filename = os.path.join(crafter.current_workspace, 'test.jpeg')
+    img_h = 20
+    img_w = 30
+    create_test_image(filename, size_width=img_w, size_height=img_h)
+    test_doc = crafter.craft(buffer=None, uri=filename, blob=None)
+    assert test_doc['blob'].shape == (img_h, img_w, 3)
 
 
 def test_io_buffer():
     crafter = ImageReader()
-    tmp_fn = os.path.join(crafter.current_workspace, 'test.jpeg')
-    img_size = 50
-    create_test_image(tmp_fn, size_width=img_size, size_height=img_size)
+    filename = os.path.join(crafter.current_workspace, 'test.jpeg')
+    img_h = 20
+    img_w = 30
+    create_test_image(filename, size_width=img_w, size_height=img_h)
     image_buffer = io.BytesIO()
-    img = Image.open(tmp_fn)
+    img = Image.open(filename)
     img.save(image_buffer, format='PNG')
     image_buffer.seek(0)
-    test_doc = crafter.craft(buffer=image_buffer.getvalue(), uri=None)
-    assert test_doc['blob'].shape == (img_size, img_size, 3)
+    test_doc = crafter.craft(buffer=image_buffer.getvalue(), uri=None, blob=None)
+    assert test_doc['blob'].shape == (img_h, img_w, 3)
     np.testing.assert_almost_equal(test_doc['blob'], np.array(img).astype('float32'))
+
+
+def test_io_blob():
+    crafter = ImageReader()
+    img_h = 20
+    img_w = 30
+    numpy_image = np.random.randint(0, 255, size=(img_h, img_w, 3), dtype=np.uint8)
+    test_doc = crafter.craft(buffer=None, uri=None, blob=numpy_image)
+    assert test_doc['blob'].shape == (img_h, img_w, 3)
+
+
+def test_blob_no_color_channel():
+    crafter = ImageReader()
+    img_h = 20
+    img_w = 30
+    numpy_image = np.random.randint(0, 255, size=(img_h, img_w), dtype=np.uint8)
+    test_doc = crafter.craft(buffer=None, uri=None, blob=numpy_image)
+    assert test_doc['blob'].shape == (img_h, img_w, 3)
```

Comment:
why does the `result` have 3 color channels then?

Reply:
The result is intended to be a standard format for images. Our image crafters assume a color channel dimension and the encoders like BigTransfer require ndarrays in RGB as well.
For instance, if you use the mnist dataset from keras,
`(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()`
the image data does not contain a color channel. 
In the current implementation, an application developer has to introduce the color channel first.
If we have encoders in the future which uses just one channel or even an alpha channel, than we could provide an output mode in the the loders constructor (RGB, RGBA, GRAYSCALE...)

Reply:
To me this is an assumption tht someone would use these encoders, while every executor should be agnostic and independent to what others do. It is strange to me that a Reader class actually transforms so deeply my input.